File: crates/vision-calibration/examples/handeye_session.rs
//! Single-camera hand-eye calibration with real KUKA robot data.
//!
//! This example demonstrates the full hand-eye calibration workflow:
//! 1. Load images and robot poses from the KUKA dataset
//! 2. Detect chessboard corners in each image
//! 3. Initialize intrinsics (Zhang's method + distortion)
//! 4. Optimize intrinsics (bundle adjustment)
//! 5. Initialize hand-eye transform (Tsai-Lenz)
//! 6. Optimize hand-eye (bundle adjustment with robot pose priors)
//!
//! Dataset: Uses `data/kuka_1/` containing:
//! - 30 images (01.png to 30.png)
//! - Robot poses (RobotPosesVec.txt)
//! - Board square size (squaresize.txt)
//!
//! Run with: `cargo run -p vision-calibration --example handeye_session`

#[path = "support/handeye_io.rs"]
mod handeye_io;

use anyhow::Result;
use chess_corners::ChessConfig;
use std::io::{self, Write};
use std::path::PathBuf;
use vision_calibration::prelude::*;
use vision_calibration::single_cam_handeye::{
    HandeyeMeta, SingleCamHandeyeExport, SingleCamHandeyeInput, SingleCamHandeyeProblem,
    SingleCamHandeyeView, step_handeye_init, step_handeye_optimize, step_intrinsics_init,
    step_intrinsics_optimize,
};

use crate::handeye_io::DatasetSummary;

fn main() -> Result<()> {
    println!("=== Single-Camera Hand-Eye Calibration (KUKA Dataset) ===\n");

    // Dataset path
    let repo_root = PathBuf::from(env!("CARGO_MANIFEST_DIR")).join("../..");
    let data_path = repo_root.join("data/kuka_1");

    if !data_path.exists() {
        anyhow::bail!(
            "Dataset not found at {}\nPlease ensure the KUKA dataset is available.",
            data_path.display()
        );
    }

    println!("Dataset: {}", data_path.display());

    // Load dataset with corner detection
    println!("\nLoading images and detecting corners...");
    let chess_config = ChessConfig::default();
    let board_params = handeye_io::kuka_chessboard_params();

    let (samples, summary) = handeye_io::load_kuka_dataset_with_progress(
        &data_path,
        &chess_config,
        &board_params,
        |current, total| {
            print!("\r  Processing image {current}/{total}...");
            io::stdout().flush().ok();
        },
    )?;

    print_dataset_summary(&summary);

    if samples.len() < 3 {
        anyhow::bail!("Need at least 3 valid views, got {}", samples.len());
    }

    // Convert to SingleCamHandeyeInput
    let views: Vec<SingleCamHandeyeView> = samples
        .into_iter()
        .map(|s| SingleCamHandeyeView {
            obs: s.view,
            meta: HandeyeMeta {
                base_se3_gripper: s.robot_pose,
            },
        })
        .collect();

    let input = SingleCamHandeyeInput::new(views)?;
    println!("Created input with {} views\n", input.num_views());

    // Create calibration session
    let mut session = CalibrationSession::<SingleCamHandeyeProblem>::new();
    session.set_input(input)?;

    // Step 1: Intrinsics initialization
    println!("--- Step 1: Intrinsics Initialization ---");
    step_intrinsics_init(&mut session, None)?;

    let init_cam = session.state.initial_camera.as_ref().unwrap();
    println!(
        "  Intrinsics: fx={:.1}, fy={:.1}, cx={:.1}, cy={:.1}",
        init_cam.k.fx, init_cam.k.fy, init_cam.k.cx, init_cam.k.cy
    );
    println!(
        "  Distortion: k1={:.4}, k2={:.4}, p1={:.5}, p2={:.5}",
        init_cam.dist.k1, init_cam.dist.k2, init_cam.dist.p1, init_cam.dist.p2
    );
    println!();

    // Step 2: Intrinsics optimization
    println!("--- Step 2: Intrinsics Optimization ---");
    step_intrinsics_optimize(&mut session, None)?;

    let opt_cam = session.state.optimized_camera.as_ref().unwrap();
    println!(
        "  Intrinsics: fx={:.1}, fy={:.1}, cx={:.1}, cy={:.1}",
        opt_cam.k.fx, opt_cam.k.fy, opt_cam.k.cx, opt_cam.k.cy
    );
    println!(
        "  Distortion: k1={:.4}, k2={:.4}, p1={:.5}, p2={:.5}",
        opt_cam.dist.k1, opt_cam.dist.k2, opt_cam.dist.p1, opt_cam.dist.p2
    );
    let reproj_err = session.state.intrinsics_reproj_error.unwrap_or(f64::NAN);
    println!("  Reprojection error: {:.4} px", reproj_err);
    println!();

    // Step 3: Hand-eye initialization
    println!("--- Step 3: Hand-Eye Initialization (Tsai-Lenz) ---");
    step_handeye_init(&mut session, None)?;

    let init_he = session.state.initial_gripper_se3_camera.as_ref().unwrap();
    let init_target = session.state.initial_base_se3_target.as_ref().unwrap();
    println!("  Hand-eye |t|: {:.4}m", init_he.translation.vector.norm());
    println!(
        "  Target in base |t|: {:.4}m",
        init_target.translation.vector.norm()
    );
    println!();

    // Step 4: Hand-eye optimization
    println!("--- Step 4: Hand-Eye Optimization ---");
    step_handeye_optimize(&mut session, None)?;

    let reproj_err = session.state.handeye_reproj_error.unwrap_or(f64::NAN);
    println!("  Final reprojection error: {:.4} px", reproj_err);
    println!();

    // Export and display final results
    println!("--- Final Results ---");
    let export = session.export()?;
    print_final_result(&export);

    Ok(())
}

fn print_final_result(export: &SingleCamHandeyeExport) {
    println!("Camera intrinsics:");
    println!("  fx = {:.2}", export.camera.k.fx);
    println!("  fy = {:.2}", export.camera.k.fy);
    println!("  cx = {:.2}", export.camera.k.cx);
    println!("  cy = {:.2}", export.camera.k.cy);

    println!("Distortion:");
    println!("  k1 = {:.6}", export.camera.dist.k1);
    println!("  k2 = {:.6}", export.camera.dist.k2);
    println!("  k3 = {:.6}", export.camera.dist.k3);
    println!("  p1 = {:.6}", export.camera.dist.p1);
    println!("  p2 = {:.6}", export.camera.dist.p2);

    match export.handeye_mode {
        HandEyeMode::EyeInHand => {
            let he = export
                .gripper_se3_camera
                .expect("missing gripper_se3_camera");
            println!("Hand-eye transform (gripper → camera):");
            let he_t = he.translation.vector;
            let he_q = he.rotation;
            println!(
                "  Translation: [{:.4}, {:.4}, {:.4}]m",
                he_t.x, he_t.y, he_t.z
            );
            println!(
                "  Rotation (quat): [{:.4}, {:.4}, {:.4}, {:.4}]",
                he_q.i, he_q.j, he_q.k, he_q.w
            );

            let target = export.base_se3_target.expect("missing base_se3_target");
            println!("Target pose in base frame:");
            let target_t = target.translation.vector;
            println!(
                "  Translation: [{:.4}, {:.4}, {:.4}]m",
                target_t.x, target_t.y, target_t.z
            );
        }
        HandEyeMode::EyeToHand => {
            let he = export.camera_se3_base.expect("missing camera_se3_base");
            println!("Hand-eye transform (camera → base):");
            let he_t = he.translation.vector;
            let he_q = he.rotation;
            println!(
                "  Translation: [{:.4}, {:.4}, {:.4}]m",
                he_t.x, he_t.y, he_t.z
            );
            println!(
                "  Rotation (quat): [{:.4}, {:.4}, {:.4}, {:.4}]",
                he_q.i, he_q.j, he_q.k, he_q.w
            );

            let target = export
                .gripper_se3_target
                .expect("missing gripper_se3_target");
            println!("Target pose in gripper frame:");
            let target_t = target.translation.vector;
            println!(
                "  Translation: [{:.4}, {:.4}, {:.4}]m",
                target_t.x, target_t.y, target_t.z
            );
        }
    }

    println!("Reprojection error:");
    println!("  Mean: {:.4} px", export.mean_reproj_error);
    println!("  Per-camera: {:?}", export.per_cam_reproj_errors);

    if let Some(deltas) = &export.robot_deltas {
        println!("Robot pose refinements ({} views):", deltas.len());
        let max_rot: f64 = deltas
            .iter()
            .map(|d| (d[0] * d[0] + d[1] * d[1] + d[2] * d[2]).sqrt())
            .max_by(|a, b| a.partial_cmp(b).unwrap())
            .unwrap_or(0.0);
        let max_trans: f64 = deltas
            .iter()
            .map(|d| (d[3] * d[3] + d[4] * d[4] + d[5] * d[5]).sqrt())
            .max_by(|a, b| a.partial_cmp(b).unwrap())
            .unwrap_or(0.0);
        println!(
            "  Max rotation delta: {:.4} rad ({:.2}°)",
            max_rot,
            max_rot.to_degrees()
        );
        println!(
            "  Max translation delta: {:.4}m ({:.2}mm)",
            max_trans,
            max_trans * 1000.0
        );
    }
}

fn print_dataset_summary(summary: &DatasetSummary) {
    println!(
        "\r  Processed {} images                ",
        summary.total_images
    );
    println!(
        "  Board: 17x28, square size {:.1}mm",
        summary.square_size_m * 1000.0
    );
    println!(
        "  Valid views: {} (skipped: {})",
        summary.used_views, summary.skipped_views
    );
    println!("  Total corners: {}", summary.total_corners);
    println!();
}

File: crates/vision-calibration-py/examples/handeye_session.py
"""Single-camera hand-eye calibration with real KUKA dataset images.

Python counterpart of:
`crates/vision-calibration/examples/handeye_session.rs`
"""

from __future__ import annotations

from pathlib import Path

import calib_targets
import numpy as np
from PIL import Image
import vision_calibration as vc

from _common import quat_xyzw_from_rot, repo_root

BOARD_ROWS = 17
BOARD_COLS = 28


def load_gray(path: Path) -> np.ndarray:
    return np.asarray(Image.open(path).convert("L"), dtype=np.uint8)


def parse_square_size_m(path: Path) -> float:
    raw = path.read_text(encoding="utf-8").strip().lower()
    if raw.endswith("mm"):
        return float(raw[:-2].strip()) / 1000.0
    if raw.endswith("m"):
        return float(raw[:-1].strip())
    value = float(raw)
    return value / 1000.0 if value > 1.0 else value


def parse_robot_poses(path: Path) -> list[vc.Pose]:
    poses: list[vc.Pose] = []
    for line in path.read_text(encoding="utf-8").splitlines():
        line = line.strip()
        if not line:
            continue
        values = [float(v) for v in line.split()]
        if len(values) != 16:
            raise ValueError(f"expected 16 values per pose line, got {len(values)}")
        r = [
            [values[0], values[1], values[2]],
            [values[4], values[5], values[6]],
            [values[8], values[9], values[10]],
        ]
        t = (values[3], values[7], values[11])
        poses.append(vc.Pose(rotation_xyzw=quat_xyzw_from_rot(r), translation_xyz=t))
    return poses


def detect_observation(path: Path, square_size_m: float) -> vc.Observation | None:
    image = load_gray(path)
    detection = calib_targets.detect_chessboard(
        image,
        chess_cfg=calib_targets.ChessConfig(),
        params=calib_targets.ChessboardParams(
            expected_rows=BOARD_ROWS,
            expected_cols=BOARD_COLS,
        ),
    )
    if detection is None:
        return None

    points_3d: list[tuple[float, float, float]] = []
    points_2d: list[tuple[float, float]] = []
    for corner in detection.detection.corners:
        if corner.grid is None:
            continue
        points_3d.append((float(corner.grid.i) * square_size_m, float(corner.grid.j) * square_size_m, 0.0))
        points_2d.append((float(corner.position[0]), float(corner.position[1])))
    if len(points_2d) < 4:
        return None
    return vc.Observation(points_3d=points_3d, points_2d=points_2d)


def load_dataset() -> vc.SingleCamHandeyeDataset:
    data_dir = repo_root() / "data/kuka_1"
    if not data_dir.exists():
        raise FileNotFoundError(f"dataset not found: {data_dir}")

    square_size_m = parse_square_size_m(data_dir / "squaresize.txt")
    robot_poses = parse_robot_poses(data_dir / "RobotPosesVec.txt")
    views: list[vc.SingleCamHandeyeView] = []
    skipped = 0

    for idx, robot_pose in enumerate(robot_poses, start=1):
        img_path = data_dir / f"{idx:02}.png"
        if not img_path.exists():
            skipped += 1
            continue
        obs = detect_observation(img_path, square_size_m)
        if obs is None:
            skipped += 1
            continue
        views.append(vc.SingleCamHandeyeView(observation=obs, base_se3_gripper=robot_pose))

    if len(views) < 3:
        raise RuntimeError(f"need at least 3 usable views, got {len(views)}")
    print(f"Loaded {len(views)} views ({skipped} skipped)")
    return vc.SingleCamHandeyeDataset(views=views)


def main() -> None:
    print("=== Single-Camera Hand-Eye Session (Real Images, Python) ===")
    dataset = load_dataset()
    result = vc.run_single_cam_handeye(dataset)
    cam = result.camera["k"]

    print(f"Hand-eye mode: {result.handeye_mode}")
    print(f"Mean reprojection error: {result.mean_reproj_error:.6f} px")
    print(f"Recovered camera fx={cam['fx']:.3f}, fy={cam['fy']:.3f}, cx={cam['cx']:.3f}, cy={cam['cy']:.3f}")
    if result.gripper_se3_camera is not None:
        t = result.gripper_se3_camera.translation_xyz
        print(f"Recovered gripper->camera translation: [{t[0]:.4f}, {t[1]:.4f}, {t[2]:.4f}] m")


if __name__ == "__main__":
    main()
